linera-ethereum/src/client.rs

```diff
@@ -1,112 +1,304 @@
+                // only response
+                let mut result = None;
+                // only error
+                let mut error = None;
+
+                while let Some(key) = map.next_key()? {
+                    match key {
+                        "jsonrpc" => {
+                            if jsonrpc {
+                                return Err(serde::de::Error::duplicate_field("jsonrpc"));
+                            }
+
+                            let value = map.next_value()?;
+                            if value != "2.0" {
+                                return Err(serde::de::Error::invalid_value(
+                                    serde::de::Unexpected::Str(value),
+                                    &"2.0",
+                                ));
+                            }
+
+                            jsonrpc = true;
+                        }
+                        "id" => {
+                            if id.is_some() {
+                                return Err(serde::de::Error::duplicate_field("id"));
+                            }
+
+                            let value: u64 = map.next_value()?;
+                            id = Some(value);
+                        }
+                        "result" => {
+                            if result.is_some() {
+                                return Err(serde::de::Error::duplicate_field("result"));
+                            }
+
+                            let value: &RawValue = map.next_value()?;
+                            result = Some(value);
+                        }
+                        "error" => {
+                            if error.is_some() {
+                                return Err(serde::de::Error::duplicate_field("error"));
+                            }
+
+                            let value: JsonRpcError = map.next_value()?;
+                            error = Some(value);
+                        }
+                        key => {
+                            return Err(serde::de::Error::unknown_field(
+                                key,
+                                &["id", "jsonrpc", "result", "error"],
+                            ))
+                        }
+                    }
+                }
+
+                // jsonrpc version must be present in all responses
+                if !jsonrpc {
+                    return Err(serde::de::Error::missing_field("jsonrpc"));
+                }
+
+                match (id, result, error) {
+                    (Some(id), Some(result), None) => Ok(Response::Success { id, result }),
+                    (Some(id), None, Some(error)) => Ok(Response::Error { id, error }),
+                    _ => Err(serde::de::Error::custom(
+                        "response must be either a success/error or notification object",
+                    )),
+                }
+            }
+        }
+
+        deserializer.deserialize_map(ResponseVisitor(&()))
+    }
+}
```

Comment:
Can't we just derive `Deserialize`? It seems to be using the internally tagged enum representation, which is supported by `serde_json`. And we can annotate the `RawValue` with `#[serde(borrow)]`.

We could do the additional checks (like `jsonrpc` version) after deserialization.

Reply:
Yes, I change that and now it is much simpler.